pkg/cli/admin/catalog/mirror.go

```diff
@@ -356,15 +378,136 @@ func (o *MirrorCatalogOptions) Complete(cmd *cobra.Command, args []string) error
+	return IndexExtractorFunc(func(from imagesource.TypedImageReference) (string, error) {
+		e := imgextract.NewExtractOptions(o.IOStreams)
+		e.SecurityOptions = o.SecurityOptions
+		e.FilterOptions = o.FilterOptions
+		e.ParallelOptions = o.ParallelOptions
+		e.FileDir = o.FileDir
+		if len(o.FromFileDir) > 0 {
+			e.FileDir = o.FromFileDir
+		}
+		e.Paths = []string{o.IndexPath}
+		e.Confirm = true
+		if err := e.Complete(cmd, []string{o.SourceRef.String()}); err != nil {
+			return "", err
+		}
+		if err := e.Validate(); err != nil {
+			return "", err
+		}
+		if err := e.Run(); err != nil {
+			return "", err
+		}
+		if len(e.Mappings) < 1 {
+			return "", fmt.Errorf("couldn't extract declarative configs")
+		}
+
+		fmt.Fprintf(o.IOStreams.Out, "wrote declarative configs to %s\n", e.Mappings[0].To)
+		fmt.Fprintf(o.IOStreams.Out, "using declarative configs at: %s\n", e.Mappings[0].To)
+		return e.Mappings[0].To, nil
+	})
+}
+
+type sqliteRelatedImagesParser struct{}
+
+func (_ sqliteRelatedImagesParser) Parse(file string) (map[string]struct{}, error) {
+	db, err := sqlittle.Open(file)
+	if err != nil {
+		return nil, err
 	}
-	o.DatabaseExtractor = extractor
 
-	return nil
+	// get all images
+	var images = make(map[string]struct{}, 0)
+	var errs = make([]error, 0)
+	reader := func(r sqlittle.Row) {
+		var image string
+		if err := r.Scan(&image); err != nil {
+			errs = append(errs, err)
+			return
+		}
+		if image != "" {
+			images[image] = struct{}{}
+		}
+	}
+	if err := db.Select("related_image", reader, "image"); err != nil {
+		errs = append(errs, err)
+		return nil, errors.NewAggregate(errs)
+	}
+
+	// get all bundlepaths
+	if err := db.Select("operatorbundle", reader, "bundlepath"); err != nil {
+		errs = append(errs, err)
+		return nil, errors.NewAggregate(errs)
+	}
+	return images, nil
+}
+
+type declcfgMeta struct {
+	Schema        string                `json:"schema"`
+	Image         string                `json:"image"`
+	RelatedImages []declcfgRelatedImage `json:"relatedImages,omitempty"`
+}
+
+type declcfgRelatedImage struct {
+	Name  string `json:"name"`
+	Image string `json:"image"`
+}
+
+type declcfgRelatedImagesParser struct{}
+
+func (_ declcfgRelatedImagesParser) Parse(root string) (map[string]struct{}, error) {
+	rootFS := os.DirFS(root)
+
+	matcher, err := ignore.NewMatcher(rootFS, ".indexignore")
+	if err != nil {
+		return nil, err
+	}
+
+	relatedImages := map[string]struct{}{}
+	if err := fs.WalkDir(rootFS, ".", func(path string, entry fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if entry.IsDir() || matcher.Match(path, false) {
+			return nil
+		}
+		f, err := rootFS.Open(path)
+		if err != nil {
+			return err
+		}
+		defer f.Close()
+		dec := utilyaml.NewYAMLOrJSONDecoder(f, 4096)
+		for {
+			var blob declcfgMeta
+			if err := dec.Decode(&blob); err != nil {
+				if err == io.EOF {
+					break
+				}
+				return err
+			}
+			if blob.Schema != "olm.bundle" {
```

Comment:
we might get some nice future-proofing by just checking for `.Image` or `.RelatedImages` and adding if they exist (so we don't have to update this when we have `olm.bundle.v2` or `olm.staticbundles` etc)